IdentityCore/src/cache/metadata/accountMetadata/MSIDAccountMetadataCacheItem.m

```diff
@@ -0,0 +1,164 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDAccountMetadataCacheItem.h"
+#import "MSIDAccountIdentifier.h"
+#import "MSIDAuthority.h"
+#import "MSIDAuthorityFactory.h"
+
+static const NSString *AccountMetadataURLMapKey = @"URLMap";
+
+@implementation MSIDAccountMetadataCacheItem
+
+- (instancetype)initWithHomeAccountId:(NSString *)homeAccountId
+                             clientId:(NSString *)clientId
+
+{
+    if (!homeAccountId || !clientId) return nil;
+    
+    self = [super init];
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _clientId = clientId;
+        _internalMap = [NSMutableDictionary new];
+    }
+    return self;
+}
+
+#pragma mark - URL caching
+- (BOOL)setCachedURL:(NSURL *)cachedURL forRequestURL:(NSURL *)requestURL error:(NSError **)error
+{
+    if ([NSString msidIsStringNilOrBlank:cachedURL.absoluteString]
+        || [NSString msidIsStringNilOrBlank:requestURL.absoluteString])
+    {
+        if (error) *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"Either a target or request URL produces a nil string", nil, nil, nil, nil, nil);
+        
+        return NO;
+    }
+    
+    NSMutableDictionary *urlMap = _internalMap[MSID_ACCOUNT_CACHE_KEY];
+    if (!urlMap)
+    {
+        urlMap = [NSMutableDictionary new];
+        _internalMap[AccountMetadataURLMapKey] = urlMap;
+    }
+    
+    urlMap[requestURL.absoluteString] = cachedURL.absoluteString;
+    return YES;
+}
+
+- (NSURL *)cachedURL:(NSURL *)cachedURL
+{
+    NSDictionary *urlMap = _internalMap[AccountMetadataURLMapKey];
+    NSString *cachedURLString = urlMap[cachedURL.absoluteString];
+    
+    return cachedURLString ? [[NSURL alloc] initWithString:urlMap[cachedURL.absoluteString]] : nil;
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    if (!json)
+    {
+        MSID_LOG_WARN(nil, @"Tried to decode an authority map item from nil json");
+        return nil;
+    }
+    
+    self->_clientId = json[MSID_CLIENT_ID_CACHE_KEY];
+    self->_homeAccountId = json[MSID_HOME_ACCOUNT_ID_CACHE_KEY];
+
+    _internalMap = [NSMutableDictionary new];
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    dictionary[MSID_CLIENT_ID_CACHE_KEY] = self.clientId;
+    dictionary[MSID_HOME_ACCOUNT_ID_CACHE_KEY] = self.homeAccountId;
+    dictionary[MSID_ACCOUNT_CACHE_KEY] = _internalMap;
```

Comment:
Should it be _internalMap[MSID_ACCOUNT_CACHE_KEY] instead?

Reply:
This is fine, the structure is
{ MSID_ACCOUNT_CACHE_KEY:   {    AccountMetadataURLMapKey  :   { // URL Mapping Dict} }

So, per account, there can be url mapping. complicated and nested, but gives us flexibility to add more account meta data other than URL mapping going forward should we need to do so,